cs4660/graph/graph.py

```diff
@@ -0,0 +1,241 @@
+                return True
+        return False
+
+    def neighbors(self, node):
+        results = []
+        for edge in self.adjacency_list[node]:
+            results.append(edge.to_node)
+        return results
+
+    def add_node(self, node):
+        if node not in self.adjacency_list:
+            self.adjacency_list[node] = []
+            return True
+        else: return False
+
+    def remove_node(self, node):
+        if node in self.adjacency_list:
+            del self.adjacency_list[node]
+            for nodes, edges in self.adjacency_list.items():
+                    """SINGLE LINE SOLUTION"""
+                    self.adjacency_list[nodes] = [edge for edge in edges if edge.to_node != node]
+                    """ LONGER SOLUTION
+                    newlist = []
+                    for edge in edges:
+                        if edge.to_node != node:
+                            newlist.append(edge)
+                    self.adjacency_list[nodes] = newlist
+                    """
+            return True
+        else: return False
+
+    def add_edge(self, edge):
+        neighbors = self.adjacency_list[edge.from_node]
+        if edge not in neighbors:
+            neighbors.append(edge)
+            return True
+        else: return False
+
+    def remove_edge(self, edge):
+        neighbors = self.adjacency_list[edge.from_node]
+        if edge in neighbors:
+            neighbors.remove(edge)
+            return True
+        else: return False
+
+class AdjacencyMatrix(object):
+    def __init__(self):
+        # adjacency_matrix should be a two dimensions array of numbers that
+        # represents how one node connects to another
+        self.adjacency_matrix = []
+        # in additional to the matrix, you will also need to store a list of Nodes
+        # as separate list of nodes
+        self.nodes = []
+
+    def adjacent(self, node_1, node_2):
+        for edge in self.adjacency_matrix[self.__get_node_index(node_1)]:
+            if edge.to_node == node_2:
+                return True
+        return False
+
+    def neighbors(self, node):
+        results = []
+        for edge in self.adjacency_matrix[self.__get_node_index(node)]:
+            results.append(edge.to_node)
+        return results
+
+    def add_node(self, node):
+        if node not in self.nodes:
+            self.nodes.append(node)
+            self.adjacency_matrix.append([])
+            return True
+        else: return False
+
+    def remove_node(self, node):
+        if node in self.nodes:
+            index = self.__get_node_index(node)
+            del self.nodes[index]
+            return True
+        else: return False
+        """
+        TODO: remove edges that attach to deleted node
+        """
+          
+
+    def add_edge(self, edge):
+        neighbors = self.adjacency_matrix[self.__get_node_index(edge.from_node)]
+        if edge not in neighbors:
+            neighbors.append(edge)
```

Comment:
this isn't really adjacency_matrix implementation. You have a data structure of the list of list of edges rather than the two dimensional array of numbers.

Reply:
e.g.

```
[
  [Edge(1, 2, 1)],
  [ ],
  [ ]
]
```

vs

```
[
  [0, 1, 0],
  [0, 0, 1],
  [0, 0, 0]
]
```